src/python/pants/backend/javascript/nodejs_project.py

```diff
@@ -0,0 +1,158 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import os.path
+from dataclasses import dataclass, replace
+from pathlib import PurePath
+from typing import Iterable
+
+from pants.backend.javascript import package_json
+from pants.backend.javascript.package_json import AllPackageJson, PackageJson
+from pants.engine.collection import Collection, DeduplicatedCollection
+from pants.engine.internals.native_engine import Digest, MergeDigests
+from pants.engine.rules import Rule, collect_rules, rule
+from pants.engine.unions import UnionRule
+from pants.util.ordered_set import FrozenOrderedSet
+from pants.util.strutil import softwrap
+
+
+@dataclass(frozen=True)
+class PackageJsonWorkspace:
+    """Exists to not invalidate AllNodeJsProjects when irrelevant parts of PackageJson has
+    changed."""
+
+    name: str
+    root_dir: str
+    digest: Digest
+    workspaces: tuple[str, ...]
+
+    @classmethod
+    def from_package_json(cls, pkg_json: PackageJson) -> PackageJsonWorkspace:
+        return PackageJsonWorkspace(
+            name=pkg_json.name,
+            root_dir=pkg_json.root_dir,
+            digest=pkg_json.digest,
+            workspaces=pkg_json.workspaces,
+        )
+
+
+class PackageJsonWorkspaces(DeduplicatedCollection[PackageJsonWorkspace]):
+    pass
+
+
+@dataclass(frozen=True)
+class NodeJSProject:
+    root_dir: str
+    workspaces: FrozenOrderedSet[PackageJsonWorkspace]
+
+    def is_parent(self, project: NodeJSProject) -> bool:
+        return self.root_dir != project.root_dir and any(
+            project.root_dir == workspace.root_dir for workspace in self.workspaces
+        )
+
+    def including_workspaces_from(self, child: NodeJSProject) -> NodeJSProject:
+        return replace(self, workspaces=self.workspaces | child.workspaces)
+
+    @property
+    def resolve_name(self) -> str:
+        return self.root_dir.replace(os.path.sep, ".")
```

Comment:
So, this is the name that has to be passed in `./pants generate-lockfiles --resolve=<name>`?

Would it make sense to allow overriding the resolve_name on the `package_json` target? Then a shorter name (probably the JS package name) could be used.

Could we come up with a more succinct, but un-surprising way to generate this resolve name?

Reply:
> Would it make sense to allow overriding the resolve_name on the package_json target? Then a shorter name (probably the JS package name) could be used.

Unfortunately not. While there is guarantee that there is a package.json for each lockfile for npm, other package managers I intend to support do not.

I guess we could allow a mapping of `path-to-lockfile: user-chosen-name` in pants.toml?

Reply:
Other ecosystems (looking at Python now as it's the one I know best) has the mapping of lockfile to resolve name in the `pants.toml` configuration so I think could make sense to have the same here. Something like
```toml
[javascript.resolves]
proj-a = "src/js/proj/a/package.json"
```

https://www.pantsbuild.org/docs/reference-python#resolves

Reply:
I think it is unnecessary to require users to assign this; meaning the current behaviour as a default is, imo, in line with the "pants not getting in your way".

But for aliasing purposes it makes sense, the lockfile path could be nested deep in some cases, spawning a particularily long dotted string.

Reply:
I'm fine with a sane default and option to override.